src/ComServerHelpers/Internal/Windows/ComPtr.cs

```diff
@@ -0,0 +1,287 @@
+    /// </summary>
+    /// <param name="ptr">The target raw pointer to copy the address of the current COM object to.</param>
+    /// <returns>The result of <see cref="IUnknown.QueryInterface(Guid*, void**)"/> for the target type <typeparamref name="U"/>.</returns>
+    public readonly HRESULT CopyTo<U>(U** ptr)
+        where U : unmanaged, IComIID
+    {
+        Guid iid = U.Guid;
+        return ((IUnknown*)ptr_)->QueryInterface(&iid, (void**)ptr);
+    }
+
+    /// <summary>
+    /// Converts the current COM object reference to a given interface type and assigns that to a target raw pointer.
+    /// </summary>
+    /// <param name="ptr">The target raw pointer to copy the address of the current COM object to.</param>
+    /// <returns>The result of <see cref="IUnknown.QueryInterface(Guid*, void**)"/> for the target type <typeparamref name="U"/>.</returns>
+    public readonly HRESULT CopyTo<U>(ref U* ptr)
+        where U : unmanaged, IComIID
+    {
+        void** tmp = null;
+
+        Guid iid = U.Guid;
+        HRESULT @ref = ((IUnknown*)ptr_)->QueryInterface(&iid, tmp);
+
+        ptr = (U*)tmp;
+
+        return @ref;
+    }
+
+    /// <summary>
+    /// Converts the current object reference to a type indicated by the given IID and assigns that to a target address.
+    /// </summary>
+    /// <param name="riid">The IID indicating the interface type to convert the COM object reference to.</param>
+    /// <param name="ptr">The target raw pointer to copy the address of the current COM object to.</param>
+    /// <returns>The result of <see cref="IUnknown.QueryInterface(Guid*, void**)"/> for the target IID.</returns>
+    public readonly int CopyTo(Guid* riid, void** ptr)
+    {
+        return ((IUnknown*)ptr_)->QueryInterface(riid, ptr);
+    }
+
+    /// <inheritdoc/>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Dispose()
+    {
+        T* pointer = ptr_;
+
+        if (pointer is not null)
+        {
+            ptr_ = null;
+
+            _ = ((IUnknown*)pointer)->Release();
+        }
+    }
+
+    /// <summary>
+    /// Gets the currently wrapped raw pointer to a COM object.
+    /// </summary>
+    /// <returns>The raw pointer wrapped by the current <see cref="ComPtr{T}"/> instance.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public readonly T* Get()
+    {
+        return ptr_;
+    }
+
+    /// <summary>
+    /// Gets the address of the current <see cref="ComPtr{T}"/> instance as a raw <typeparamref name="T"/> double pointer.
+    /// This method is only valid when the current <see cref="ComPtr{T}"/> instance is on the stack or pinned.
+    /// </summary>
+    /// <returns>The raw pointer to the current <see cref="ComPtr{T}"/> instance.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public readonly T** GetAddressOf()
+    {
+        return (T**)Unsafe.AsPointer(ref Unsafe.AsRef(in this));
```

Comment:
This is a GC hole waiting to happen. The only way this is ever correct is if the `ComPtr` is forced to always be on the stack, but there is nothing here that guarantees that. Creating an `ComPtr<>[]` makes this code invalid.

Reply:
Is this a new addition? I remember the compiler giving errors when I'm using `in this` in a ref struct.

Reply:
Updates usages and removed `ComPtr` altogether.